Enable read/write real[int,int] in binary format + fix for some compilation problem due to removal of msh3 plugin

ffeppon opened the pull request:
Hi, 

For pyfreefem support, I enhanced the bfstream.cpp plugin to enable read/write of real[int,int] array in binary. 
It would be great to have the same for the "matrix" type, but I don't know how to do it for the moment. 

I also noticed that [this commit](https://github.com/FreeFem/FreeFem-sources/commit/fc9dc8b7c7f6d52757cb2ae6dd4680d715e3e063) seems to break the compilation from my side... Below a small fix. 

Regards
Florian

frederichecht commented:
Hi, your commit break freefem++, 
because the erreur is due to bad update


Best Regards, 

Frédéric Hecht.   
   

--------------------------------------------------------------------
Laboratoire  Jacques-Louis Lions,          UPMC Sorbonne Université
BC187,  4 Place  Jussieu,  75252    PARIS cedex 05,          France
Campus Jussieu,  Barre 15-25,          3 etage           Bureau 307 
Projet Alpines,  Centre Inria de Paris, 48 rue Barrault 75013 Paris
tel:  +33 1 44274411,  mob: +33 6 62198986
***@***.***     https://www.ljll.fr/hecht
software:  FreeFem++    web site:       http://www.freefem.org/ff++





> Le 25 févr. 2025 à 09:15, Florian Feppon ***@***.***> a écrit :
> 
> Hi,
> 
> For pyfreefem support, I enhanced the bfstream.cpp plugin to enable read/write of real[int,int] array in binary.
> It would be great to have the same for the "matrix" type, but I don't know how to do it for the moment.
> 
> I also noticed that this commit <https://github.com/FreeFem/FreeFem-sources/commit/fc9dc8b7c7f6d52757cb2ae6dd4680d715e3e063> seems to break the compilation from my side... Below a small fix.
> 
> Regards
> Florian
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/FreeFem/FreeFem-sources/pull/333
> 
> Commit Summary
> 
> 1790433 <https://github.com/FreeFem/FreeFem-sources/pull/333/commits/1790433780c88922e73a0e58fc639b19d2fc5a75> Support read/write real[int,int] arrays into binary format with bfstream plugin
> 5653920 <https://github.com/FreeFem/FreeFem-sources/pull/333/commits/56539208f8b3f300f0c6e6fff68cc457d06b9a4a> Fix some compilation problems due to removal of msh3 plugin
> File Changes (7 files <https://github.com/FreeFem/FreeFem-sources/pull/333/files>)
> M plugin/seq/aniso.cpp <https://github.com/FreeFem/FreeFem-sources/pull/333/files#diff-0b4ee844b6340a2b7069a9c7762163e283e4626842f3dddbdc71bbdea9da762f> (4)
> M plugin/seq/bfstream.cpp <https://github.com/FreeFem/FreeFem-sources/pull/333/files#diff-ed5a80803cd617db93cf6f6148499e6da0229594d1e15b3fa616aab7e4634ae3> (65)
> M plugin/seq/mshmet.cpp <https://github.com/FreeFem/FreeFem-sources/pull/333/files#diff-7e07b8c81c1915d53d9774ccc3e2067ac0c633d767d5fd25edf4edf9e6c9354d> (2)
> M plugin/seq/splitmesh12.cpp <https://github.com/FreeFem/FreeFem-sources/pull/333/files#diff-c04bbde2894a73bb9a813570dbff3f2fdd0a671fcf65561a74e392a1051015c2> (2)
> M plugin/seq/splitmesh4.cpp <https://github.com/FreeFem/FreeFem-sources/pull/333/files#diff-7f3a8f3eabc85daf39c93f4d8cb92a943e80f0ba03222ddf0eacfc4182ebdd7e> (2)
> M plugin/seq/tetgen.cpp <https://github.com/FreeFem/FreeFem-sources/pull/333/files#diff-11b8f12475fee301f0e280fb2cc8869e4680457ca118e556eb095764cf7c020a> (2)
> M src/lglib/lg.tab.cpp <https://github.com/FreeFem/FreeFem-sources/pull/333/files#diff-db75ce486f1a4228440f38036340538ddf5bfe13da1772ba2ab5fb6859d3e112> (4)
> Patch Links:
> 
> https://github.com/FreeFem/FreeFem-sources/pull/333.patch
> https://github.com/FreeFem/FreeFem-sources/pull/333.diff
> —
> Reply to this email directly, view it on GitHub <https://github.com/FreeFem/FreeFem-sources/pull/333>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABLPNDGGUV2PNMDGTCT2OJT2RQRDLAVCNFSM6AAAAABXZ7CVUSVHI2DSMVQWIX3LMV43ASLTON2WKOZSHA3TONJRGI4TMNQ>.
> You are receiving this because you are subscribed to this thread.
>

prj- commented:
Please revert your `msh3` changes, things are not broken as Frédéric pointed out. For sparse matrix in binary format, you can do that through PETSc.

prj- commented:
Like this: https://github.com/FreeFem/FreeFem-sources/blob/b1e524c8ca6a9b44cb9eff780459bacfc12a3ad7/examples/hpddm/stokes-block-2d-PETSc.edp#L88
A FreeFEM matrix would first need to be wrapped in a PETSc Mat, like `Mat A(B);`
This would give you much more flexibility, since you would then be able to use the matrix directly with petsc4py or slepc4py in your Python code.

frederichecht commented:
For the sparse matrix in bin air 

in  plugin matrixmarket.cpp 

savemtx("A.mtx",A,true); //  save in binary
readmtx("A.mtx",A,true); //  save in binary


> Le 25 févr. 2025 à 09:48, Pierre Jolivet ***@***.***> a écrit :
> 
> 
> prj-
>  left a comment 
> (FreeFem/FreeFem-sources#333)
> Please revert your msh3 changes, things are not broken as Frédéric pointed out. For sparse matrix in binary format, you can do that through PETSc.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.
> 
>  <https://github.com/FreeFem/FreeFem-sources/pull/333#issuecomment-2681167861> <https://github.com/notifications/unsubscribe-auth/ABLPNDDLCBZFEV53YDUICXL2RQU7JAVCNFSM6AAAAABXZ7CVUSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOBRGE3DOOBWGE>
> 
> prj-
>  left a comment 
> (FreeFem/FreeFem-sources#333)
>  <https://github.com/FreeFem/FreeFem-sources/pull/333#issuecomment-2681167861>
> Please revert your msh3 changes, things are not broken as Frédéric pointed out. For sparse matrix in binary format, you can do that through PETSc.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/FreeFem/FreeFem-sources/pull/333#issuecomment-2681167861>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABLPNDDLCBZFEV53YDUICXL2RQU7JAVCNFSM6AAAAABXZ7CVUSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOBRGE3DOOBWGE>.
> You are receiving this because you commented.
>

ffeppon commented:
> For the sparse matrix in bin air in plugin matrixmarket.cpp savemtx("A.mtx",A,true); // save in binary readmtx("A.mtx",A,true); // save in binary
> […](#)
> Le 25 févr. 2025 à 09:48, Pierre Jolivet ***@***.***> a écrit : prj- left a comment ([FreeFem/FreeFem-sources#333](https://github.com/FreeFem/FreeFem-sources/pull/333)) Please revert your msh3 changes, things are not broken as Frédéric pointed out. For sparse matrix in binary format, you can do that through PETSc. — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you commented. <[#333 (comment)](https://github.com/FreeFem/FreeFem-sources/pull/333#issuecomment-2681167861)> <https://github.com/notifications/unsubscribe-auth/ABLPNDDLCBZFEV53YDUICXL2RQU7JAVCNFSM6AAAAABXZ7CVUSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOBRGE3DOOBWGE> prj- left a comment ([FreeFem/FreeFem-sources#333](https://github.com/FreeFem/FreeFem-sources/pull/333)) <[#333 (comment)](https://github.com/FreeFem/FreeFem-sources/pull/333#issuecomment-2681167861)> Please revert your msh3 changes, things are not broken as Frédéric pointed out. For sparse matrix in binary format, you can do that through PETSc. — Reply to this email directly, view it on GitHub <[#333 (comment)](https://github.com/FreeFem/FreeFem-sources/pull/333#issuecomment-2681167861)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABLPNDDLCBZFEV53YDUICXL2RQU7JAVCNFSM6AAAAABXZ7CVUSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOBRGE3DOOBWGE>. You are receiving this because you commented.

Thank you Frédéric! I have now implemented import/export from FreeFEM to Python using binary format (https://gitlab.com/florian.feppon/pyfreefem/-/commit/2b6e570f70c7e57cdd31e1aacdf260a124629a97) and also functions to decode/encode the PETSc format. 

Will you then accept my merge request for the real[int,int] type ? I have removed the problematic change, only bfstream.cpp should be modified. 

Thanks a lot 
Florian

frederichecht commented:
Yes for bstream no problem I accept. 

> Le 25 févr. 2025 à 10:51, Pierre Jolivet ***@***.***> a écrit :
> 
> 
> prj-
>  left a comment 
> (FreeFem/FreeFem-sources#333)
> Like this: https://github.com/FreeFem/FreeFem-sources/blob/b1e524c8ca6a9b44cb9eff780459bacfc12a3ad7/examples/hpddm/stokes-block-2d-PETSc.edp#L88
> A FreeFEM matrix would first need to be wrapped in a PETSc Mat, like Mat A(B);
> This would give you much more flexibility, since you would then be able to use the matrix directly with petsc4py or slepc4py in your Python code.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.
> 
>  <https://github.com/FreeFem/FreeFem-sources/blob/b1e524c8ca6a9b44cb9eff780459bacfc12a3ad7/examples/hpddm/stokes-block-2d-PETSc.edp#L88> <https://github.com/FreeFem/FreeFem-sources/pull/333#issuecomment-2681378935> <https://github.com/notifications/unsubscribe-auth/ABLPNDEQTVRLCQEJRZC427D2RQ4LJAVCNFSM6AAAAABXZ7CVUSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOBRGM3TQOJTGU>
> 
> prj-
>  left a comment 
> (FreeFem/FreeFem-sources#333)
>  <https://github.com/FreeFem/FreeFem-sources/pull/333#issuecomment-2681378935>
> Like this: https://github.com/FreeFem/FreeFem-sources/blob/b1e524c8ca6a9b44cb9eff780459bacfc12a3ad7/examples/hpddm/stokes-block-2d-PETSc.edp#L88
> A FreeFEM matrix would first need to be wrapped in a PETSc Mat, like Mat A(B);
> This would give you much more flexibility, since you would then be able to use the matrix directly with petsc4py or slepc4py in your Python code.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/FreeFem/FreeFem-sources/pull/333#issuecomment-2681378935>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABLPNDEQTVRLCQEJRZC427D2RQ4LJAVCNFSM6AAAAABXZ7CVUSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOBRGM3TQOJTGU>.
> You are receiving this because you commented.
>

ffeppon commented:
I have a case where the binary export does not work, see below. On my machine, no file "A.mtx" is saved. 

```freefem
load "MatrixMarket" 

mesh Th=square(2,2);   
    
int[int] labPeriodic = [2,4]; 
macro Pkv() [P2,P2]//
macro Pkvper() [P2,P2], periodic=[[2,y],[4,y]]//
    
fespace Fhvper(Th, Pkvper); 
fespace Fhv(Th, Pkv); 

matrix interpPvPvper = interpolate(Fhv,Fhvper); 


// Check equivalence between formulas permeability  
cout << interpPvPvper << endl; 
    
savemtx("A.mtx",interpPvPvper,true); 
```

ffeppon commented:
**Edit**: I suggest to remove the equal size check in savemtx?

https://github.com/FreeFem/FreeFem-sources/pull/333/files#diff-48ed1be82c409089e33028da3bf43e8d866614f69577ccaf6c367471f8398631R762

The change seems to have been integrated to this merge request.

frederichecht commented:
I think you must verifiy this is compatible with readmtx  and may be with other soft. 

> Le 25 févr. 2025 à 15:49, Florian Feppon ***@***.***> a écrit :
> 
> 
> ffeppon
>  left a comment 
> (FreeFem/FreeFem-sources#333)
> Edit: I suggest to remove the equal size check in savemtx? This fixes the problem above.
> 
> https://github.com/FreeFem/FreeFem-sources/pull/333/files#diff-48ed1be82c409089e33028da3bf43e8d866614f69577ccaf6c367471f8398631R762
> 
> The change seems to have been integrated to this merge request.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.
> 
>  <https://github.com/FreeFem/FreeFem-sources/pull/333/files#diff-48ed1be82c409089e33028da3bf43e8d866614f69577ccaf6c367471f8398631R762> <https://github.com/FreeFem/FreeFem-sources/pull/333#issuecomment-2682229695> <https://github.com/notifications/unsubscribe-auth/ABLPNDG4CJRXUN7LI6OC53L2RR7JBAVCNFSM6AAAAABXZ7CVUSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOBSGIZDSNRZGU>
> 
> ffeppon
>  left a comment 
> (FreeFem/FreeFem-sources#333)
>  <https://github.com/FreeFem/FreeFem-sources/pull/333#issuecomment-2682229695>
> Edit: I suggest to remove the equal size check in savemtx? This fixes the problem above.
> 
> https://github.com/FreeFem/FreeFem-sources/pull/333/files#diff-48ed1be82c409089e33028da3bf43e8d866614f69577ccaf6c367471f8398631R762
> 
> The change seems to have been integrated to this merge request.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/FreeFem/FreeFem-sources/pull/333#issuecomment-2682229695>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ABLPNDG4CJRXUN7LI6OC53L2RR7JBAVCNFSM6AAAAABXZ7CVUSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOBSGIZDSNRZGU>.
> You are receiving this because you commented.
>

ffeppon commented:
Dear Frédéric, 
I could check on my application that it works as a charm. On [this example](https://gitlab.com/florian.feppon/pyfreefem/-/blob/f5f2255a16f3390f499e11306fa6e4815db799de/pyfreefem/examples/ex18_read_write_binaryMatrix.py#L61), I save and then read a binary matrix from/to python with successive `savemtx` and `readmtx`, and it works with rectangular matrices with my fix. 

Regards

prj- commented on plugin/seq/MatrixMarket.cpp:
```diff
@@ -759,7 +759,7 @@ long savemtx(std::string *const &hb_filename, Matrice_Creuse< R > *const &sparse
     const int N = A->N;
     const int M = A->M;
-    if (N != M) return 1;
+    //if (N != M) return 1; //suspicious
```
```suggestion
```

prj- replied:
@ffeppon, could you please apply my suggestion so that we can merge this?
